From: IrfanMohammad <irfan@graphiant.com>
Date: Sat, 30 Sep 2023 22:16:28 +0000
Subject: ignore isolate flag for operational datastore

ignore child deletes after parent delete with isolate
ignore isolate for set items in oper
---
 src/sysrepo.c          | 21 +++++++++++++++++++--
 tests/test_oper_push.c | 39 +++++++++++++++++++++++++++++++++++++++
 2 files changed, 58 insertions(+), 2 deletions(-)

diff --git a/src/sysrepo.c b/src/sysrepo.c
index 87496b9..220b8ec 100644
--- a/src/sysrepo.c
+++ b/src/sysrepo.c
@@ -3308,10 +3308,16 @@ sr_free_values(sr_val_t *values, size_t count)
 }
 
 API int
-sr_set_item(sr_session_ctx_t *session, const char *path, const sr_val_t *value, const sr_edit_options_t opts)
+sr_set_item(sr_session_ctx_t *session, const char *path, const sr_val_t *value, const sr_edit_options_t in_opts)
 {
     sr_error_info_t *err_info = NULL;
     char str[22], *str_val;
+    sr_edit_options_t opts = in_opts;
+
+    if (session && session->ds == SR_DS_OPERATIONAL) {
+        /* operational datastore does not support isolate flag */
+        opts &= ~SR_EDIT_ISOLATE;
+    }
 
     SR_CHECK_ARG_APIRET(!session || (!path && (!value || !value->xpath)) || SR_EDIT_DS_API_CHECK(session->ds, opts),
             session, err_info);
@@ -3336,10 +3342,16 @@ sr_set_item(sr_session_ctx_t *session, const char *path, const sr_val_t *value,
 
 API int
 sr_set_item_str(sr_session_ctx_t *session, const char *path, const char *value, const char *origin,
-        const sr_edit_options_t opts)
+        const sr_edit_options_t in_opts)
 {
     sr_error_info_t *err_info = NULL;
     char *pref_origin = NULL;
+    sr_edit_options_t opts = in_opts;
+
+    if (session && session->ds == SR_DS_OPERATIONAL) {
+        /* operational datastore does not support isolate flag */
+        opts &= ~SR_EDIT_ISOLATE;
+    }
 
     SR_CHECK_ARG_APIRET(!session || !path || SR_EDIT_DS_API_CHECK(session->ds, opts), session, err_info);
 
@@ -3387,6 +3399,11 @@ sr_delete_item(sr_session_ctx_t *session, const char *path, const sr_edit_option
     const char *operation;
     const struct lysc_node *snode;
     uint32_t temp_lo = 0;
+    int isolate = SR_EDIT_ISOLATE & opts;
+
+    if (isolate && SR_DS_OPERATIONAL == session->ds) {
+        return sr_discard_items(session, path);
+    }
 
     SR_CHECK_ARG_APIRET(!session || !path || SR_EDIT_DS_API_CHECK(session->ds, opts), session, err_info);
 
diff --git a/tests/test_oper_push.c b/tests/test_oper_push.c
index 82819a2..eb86dd8 100644
--- a/tests/test_oper_push.c
+++ b/tests/test_oper_push.c
@@ -159,6 +159,44 @@ clear_up(void **state)
     return 0;
 }
 
+static void
+test_oper_set_del_new(void **state)
+{
+    struct state *st = (struct state *)*state;
+    int i, ret;
+
+    char *xpath_set[] = {
+        "/ietf-interfaces:interfaces-state/interface[name='eth1']/type",
+        "/ietf-interfaces:interfaces-state/interface[name='eth1']/oper-status",
+        NULL
+    };
+
+    char *xpath_set_vals[] = {
+        "iana-if-type:ethernetCsmacd",
+        "down",
+        NULL
+    };
+
+    char *xpath_del[] = {
+        "/ietf-interfaces:interfaces-state/interface[name='eth1']",
+        "/ietf-interfaces:interfaces-state/interface[name='eth1']/oper-status",
+        NULL
+    };
+
+    for (i = 0; xpath_set[i]; i++) {
+        ret = sr_set_item_str(st->sess, xpath_set[i], xpath_set_vals[i], NULL, SR_EDIT_ISOLATE);
+        assert_int_equal(ret, SR_ERR_OK);
+    }
+    assert_int_equal(ret, SR_ERR_OK);
+
+    for (i = 0; xpath_del[i]; i++) {
+        ret = sr_delete_item(st->sess, xpath_del[i], SR_EDIT_ISOLATE);
+        assert_int_equal(ret, SR_ERR_OK);
+    }
+    ret = sr_apply_changes(st->sess, 0);
+    assert_int_equal(ret, SR_ERR_OK);
+}
+
 static int
 dummy_change_cb(sr_session_ctx_t *session, uint32_t sub_id, const char *module_name, const char *xpath, sr_event_t event,
         uint32_t request_id, void *private_data)
@@ -4076,6 +4114,7 @@ int
 main(void)
 {
     const struct CMUnitTest tests[] = {
+        cmocka_unit_test_teardown(test_oper_set_del_new, clear_up),
         cmocka_unit_test_teardown(test_conn_owner1, clear_up),
         cmocka_unit_test_teardown(test_conn_owner2, clear_up),
         cmocka_unit_test_teardown(test_conn_owner_same_data, clear_up),
